mlir/include/mlir/Analysis/Presburger/Matrix.h

```diff
@@ -0,0 +1,166 @@
+  std::vector<std::vector<INT>> data;
+};
+
+template <typename INT>
+Matrix<INT>::Matrix(unsigned rows, unsigned columns)
+    : nColumns(columns), data(rows, std::vector<INT>(columns)) {}
+
+template <typename INT>
+Matrix<INT> Matrix<INT>::getIdentityMatrix(unsigned dimension) {
+  Matrix<INT> matrix(dimension, dimension);
+  for (size_t i = 0; i < dimension; ++i)
+    matrix(i, i) = 1;
+  return matrix;
+}
+
+template <typename INT>
+INT &Matrix<INT>::operator()(unsigned row, unsigned column) {
+  assert(row < getNumRows() && "Row outside of range");
+  assert(column < getNumColumns() && "Column outside of range");
+  return data[row][column];
+}
+
+template <typename INT>
+INT Matrix<INT>::operator()(unsigned row, unsigned column) const {
+  assert(row < getNumRows() && "Row outside of range");
+  assert(column < getNumColumns() && "Column outside of range");
+  return data[row][column];
+}
+
+template <typename INT> unsigned Matrix<INT>::getNumRows() const {
+  return data.size();
+}
+
+template <typename INT> unsigned Matrix<INT>::getNumColumns() const {
+  return nColumns;
+}
+
+template <typename INT>
+void Matrix<INT>::resize(unsigned newNRows, unsigned newNColumns) {
+  nColumns = newNColumns;
+  for (auto &row : data)
+    row.resize(nColumns);
+  data.resize(newNRows, std::vector<INT>(nColumns));
+}
+
+template <typename INT>
+void Matrix<INT>::swapRows(unsigned row, unsigned otherRow) {
+  assert((row < getNumRows() && otherRow < getNumRows()) &&
+         "Given row out of bounds");
+  if (row == otherRow)
+    return;
+  swap(data[row], data[otherRow]);
+}
+
+template <typename INT>
+void Matrix<INT>::swapColumns(unsigned column, unsigned otherColumn) {
+  assert((column < getNumColumns() && otherColumn < getNumColumns()) &&
+         "Given column out of bounds");
+  if (column == otherColumn)
+    return;
+  for (auto &row : data)
+    std::swap(row[column], row[otherColumn]);
+}
+
+template <typename INT>
+const std::vector<INT> Matrix<INT>::getRow(unsigned row) const {
+  return data[row];
+}
+
+template <typename INT>
+void Matrix<INT>::addToRow(unsigned sourceRow, unsigned targetRow, INT scale) {
+  if (scale == 0)
+    return;
+  for (unsigned col = 0; col < getNumColumns(); ++col)
+    data[targetRow][col] += scale * data[sourceRow][col];
+  return;
+}
+
+template <typename INT> void Matrix<INT>::print(llvm::raw_ostream &os) const {
+  os << "Dumping matrix, rows = " << getNumRows()
+               << ", columns: " << getNumColumns() << '\n';
+  os << "r/c  ";
+  for (unsigned column = 0; column < getNumColumns(); ++column)
+    os << "| " << column << " ";
+  os << '\n';
+  os << std::string(5 + getNumColumns() * 5, '-') << '\n';
+  for (unsigned row = 0; row < getNumRows(); ++row) {
+    os << row << " | ";
+    for (unsigned column = 0; column < getNumColumns(); ++column)
+      os << data[row][column] << " ";
+    os << '\n';
+  }
+}
+
+template <typename INT> void Matrix<INT>::dump() const {
+  print(llvm::errs());
```

Comment:
why in llvm::errs and not llvm::outs?

Reply:
FlatAffineConstraints dumps to llvm:errs(), so I did the same here.

Reply:
ok.